### PR TITLE
Cache beta

### LINKED ISSUE #102: [Optimization] Beta caching
Beta diversity right now is used for k-nearest neighbors. These matrices are massive for resident memory. To mitigate resident, we could instead cache only the first K nearest neighbors for every sample (e.g., 100), which would substantially reduce the amount of information we need to retain resident

### PR DESCRIPTION
**Pull request by wasade**
Fixes #102

### CODE REVIEW

File: microsetta_public_api/_io.py [hunk 1]
```diff
@@ -15,4 +56,8 @@ def _dict_of_paths_to_beta_data(dict_of_qza_paths, resource_name) -> \
                                            DistanceMatrixType,
                                            view_type=DistanceMatrix,
                                            )
+    key = list(new_resource.keys())[0]
+    dm = new_resource[key]
+    k = min(len(dm.ids) - 1, 100)
```
Comment (wasade) [inline microsetta_public_api/_io.py:61]: ...this is hardcoded at a max of 100 neighbors. It *may* be nice to push to config, but I vote another problem for another day.

File: microsetta_public_api/repo/_beta_repo.py [hunk 2]
```diff
@@ -11,31 +10,19 @@ def __init__(self, resources):
...
 
     def k_nearest(self, sample_id, metric, k=1):
         distance_matrix = self._get_resource(metric)
         if not self.exists(sample_id, metric):
             raise UnknownID(sample_id)
-        n_neighbors = len(distance_matrix.ids) - 1
+        n_neighbors = len(distance_matrix.columns)
         if k > n_neighbors:
             raise InvalidParameter(
-                f"k={k} is greater than the number of neighbors of the "
```
Comment (gwarmstrong) [inline microsetta_public_api/repo/_beta_repo.py:25]: I think it is useful to include the maximum in this error message. Can you include that?

File: microsetta_public_api/_io.py (hunk 1 above)
Comment (gwarmstrong) [inline microsetta_public_api/_io.py:61]:
```suggestion
    max_n_neighbors = 100
    k = min(len(dm.ids) - 1, max_n_neighbors)
```

**Review by gwarmstrong — CHANGES_REQUESTED** [review]
Looks good! Just a couple minor comments.

File: microsetta_public_api/repo/_beta_repo.py (hunk 2 above)
Comment (wasade) [reply microsetta_public_api/repo/_beta_repo.py:25]: yes!

**Comment by wasade** [comment]
Thanks!! so we'll still have long load times but resident will be small so we can spin up many servers to handle requests. 

eventually we should do this directly from the hdf5 distance matrices which should greatly improve start up time, but that is more complex

**Comment by gwarmstrong** [comment]
How complex would it be to add a `k_neighbors` to the results pipeline and save the resulting sparse distance matrix as a biom table? Loading biom is already supported by the public API and converting to something like the DataFrame in this PR should not be too hard.

**Comment by wasade** [comment]
Great idea. It can't be biom though as the table contains str not int or float. 

Could you knock out the config pieces for this? i can produce a tabular file suitable for `pd.from_csv('foo', sep='\t', dtype=str).set_index('sample-id')` where the columns are like what's here. I'll migrate this neighbor code here into microsetta-processing, PR in ~15 i think

There are multiple benefits: load time, resident memory, and we can more easily control what set of samples the neighbors are too (e.g., TMI sample, with neighbors assessed via non-human vertebrates)

**Comment by gwarmstrong** [comment]
>  It can't be biom though as the table contains str not int or float.

so the idea was something approximately like
```
# in processing
from biom import Table
table = Table(sparse_distance_matrix, sample_ids, sample_ids)

# in public api
spare_dm = read_biom('./path/to/table.biom')
cached_ids = create_pandas_cache(spare_dm)
```

but the tabular file could work too

**Comment by wasade** [comment]
Huh, hadn't thought of it like that. I don't _think_ we have a use case right now for the actual distances?

**Comment by gwarmstrong** [comment]
No as far as I’m aware

**Comment by wasade** [comment]
Okay, code is moved to processing. I'll kcik stuff off to run over the next 24h or so so we have new files / configs

**Comment by wasade** [comment]
@gwarmstrong i think this is all hooked up now to the cached datasource

**Review by gwarmstrong — COMMENTED** [review]
Looks good! A couple minor comments. I think its worth discussing whether the `BetaRepo` code should be retained even if there is not currently a use for it right now.

**Comment by wasade** [comment]
Thanks @gwarmstrong! I'm testing this branch right now and there is still an unexplained massive memory use, so gathering a little more information right now. I'll circle back to the comments as soon as I can

File: microsetta_public_api/api/diversity/beta.py [hunk 3]
```diff
@@ -1,25 +1,24 @@
...
-
-    beta_kw = schema.beta_kw
-    if not dataset_resource.has(beta_kw):
-        raise UnknownResource(f"No beta data (kw: '{beta_kw}') for "
-                              f"dataset='{dataset}'.")
-    beta_resource = dataset_resource.gets(beta_kw)
-    return beta_resource
+    try:
+        neighbors_resource = dataset_resource.gets(schema.neighbors_kw)
+    except KeyError:
```
[the remark this answers is not in the document]
Comment (wasade) [reply microsetta_public_api/api/diversity/beta.py:17]: TBH, I was just adapting the validate alpha code. It's favorable if you expect it to happen regularly. Exceptions are expensive. Changing here as a 404 wouldn't be necessarily unusual (if other people are using the API)

File: microsetta_public_api/api/tests/test_integration.py [hunk 4]
```diff
@@ -1436,28 +1435,24 @@ class BetaIntegrationTests(IntegrationTests):
...
+                                       ['s3', 's1', 's2']],
+                                      columns=['k0', 'k1', 'k2'],
+                                      index=['s1', 's2', 's3', 's4'])
+        self.neighbors.index.name = 'sample_id'
+
+        self.neighbors_path = self.create_tempfile(suffix='.tsv').name
+        self.neighbors.to_csv(self.neighbors_path, sep='\t', index=True,
+                              header=True)
+        import os
+        print(os.stat(self.neighbors_path))
```
[the remark this answers is not in the document]
Comment (wasade) [reply microsetta_public_api/api/tests/test_integration.py:1450]: Good spot

File: microsetta_public_api/api/diversity/beta.py [hunk 5]
```diff
@@ -1,25 +1,24 @@
...
     UnknownResource,
 )
 from microsetta_public_api.config import schema
 
 
-def _validate_dataset_beta(dataset, resource_getter):
-    getter = resource_getter()
-    dataset_info = ('datasets', dataset)
-    if not getter.has(*dataset_info):
+def _validate_dataset_neighbors(dataset, resource_getter):
```
[the remark this answers is not in the document]
Comment (wasade) [reply microsetta_public_api/api/diversity/beta.py:10]: It's in the github history so we could just copy it if needed?

File: microsetta_public_api/repo/_beta_repo.py [hunk 6]
```diff
@@ -1,41 +1,28 @@
...
         distance_matrix = self._get_resource(metric)
         if isinstance(sample_ids, str):
-            return sample_ids in distance_matrix
+            return sample_ids in distance_matrix.index
         else:
-            return [(id_ in distance_matrix) for id_ in sample_ids]
+            return [(id_ in distance_matrix.index) for id_ in sample_ids]
 
     def k_nearest(self, sample_id, metric, k=1):
         distance_matrix = self._get_resource(metric)
```
[the remark this answers is not in the document]
Comment (wasade) [reply microsetta_public_api/repo/_beta_repo.py]: Done

File: microsetta_public_api/api/diversity/beta.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (gwarmstrong) [reply microsetta_public_api/api/diversity/beta.py:10]: Sure.

**Review by gwarmstrong — APPROVED** [review]
Looks good to me. Good to merge?

**Comment by wasade** [comment]
+1

**Pull request merged**